File: CS/NonPersistentObjectsDemo.Module/DataStoreMapper.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using DevExpress.Data.Filtering;
using DevExpress.Data.Filtering.Helpers;
using DevExpress.Xpo.DB;

namespace NonPersistentObjectsDemo.Module {

    public class DataStoreMapping {
        public DBTable Table;
        public Func<object> Create;
        public Action<object, object[], ObjectMap> Load;
        public Action<object, object[]> Save;
        public Action<object, object> SetKey;
        public Func<object, object> GetKey;
        public IEnumerable<Column> RefColumns;
        public struct Column {
            public int Index;
            public Type Type;
        }
    }

    class DataStoreObjectLoader {
        struct PreResult {
            public DataStoreMapping Mapping;
            public Type ObjectType;
            public SelectStatement Statement;
        }
        struct PostResult {
            public DataStoreMapping Mapping;
            public List<object> Objects;
            public SelectStatementResult Result;
        }
        private ObjectMap objectMap;
        private IDataStore dataStore;
        private Dictionary<Type, DataStoreMapping> mappings;
        public DataStoreObjectLoader(Dictionary<Type, DataStoreMapping> mappings, IDataStore dataStore, ObjectMap objectMap) {
            this.mappings = mappings;
            this.dataStore = dataStore;
            this.objectMap = objectMap;
        }
        private DBColumn GetKeyColumn(DataStoreMapping mapping) {
            return mapping.Table.Columns.First(c => c.IsKey);
        }
        private CriteriaOperator BuildByKeyCriteria(Type objectType, object key, string alias) {
            return new BinaryOperator(
                new QueryOperand(GetKeyColumn(mappings[objectType]), alias),
                new OperandValue(key),
                BinaryOperatorType.Equal);
        }
        private SelectStatement PhaseOne(DataStoreMapping mapping, Type objectType, CriteriaOperator dbCriteria, string alias) {
            var statement = new SelectStatement(mapping.Table, alias);
            statement.Condition = dbCriteria;
            foreach(var column in mapping.Table.Columns) {
                statement.Operands.Add(new QueryOperand(column, alias));
            }
            return statement;
        }
        private List<object> PhaseTwo(DataStoreMapping mapping, Type objectType, SelectStatementResult result, List<PreResult> toLoad) {
            List<object> objects = new List<object>();
            int keyColumnIndex = mapping.Table.Columns.IndexOf(GetKeyColumn(mapping));
            List<DataStoreMapping.Column> refColumns = mapping.RefColumns.ToList();
            foreach(var row in result.Rows) {
                var key = row.Values[keyColumnIndex];
                if(key == null)
                    throw new DataException("Key cannot be null.");
                var obj = objectMap.Get(objectType, key);
                if(obj == null) {
                    obj = mapping.Create();
                    objectMap.Add(objectType, key, obj);
                }
                objects.Add(obj);
                foreach(var member in refColumns) {
                    var alias = "T";
                    var dbCriteria = BuildByKeyCriteria(member.Type, row.Values[member.Index], alias);
                    var memberMapping = mappings[member.Type];
                    toLoad.Add(new PreResult() {
                        Mapping = memberMapping,
                        ObjectType = member.Type,
                        Statement = PhaseOne(memberMapping, member.Type, dbCriteria, alias)
                    });
                }
            }
            return objects;
        }
        private void PhaseThree(DataStoreMapping mapping, List<object> objects, SelectStatementResult result) {
            for(int i = 0; i < objects.Count; i++) {
                mapping.Load(objects[i], result.Rows[i].Values, objectMap);
            }
        }
        public IList<object> LoadObjects(Type objectType, CriteriaOperator criteria) {
            var mapping = mappings[objectType];
            var alias = "T";
            var dbCriteria = SimpleDataStoreCriteriaVisitor.Transform(criteria, mapping.Table, alias);
            return LoadObjectsCore(objectType, dbCriteria, alias);
        }
        private IList<object> LoadObjectsCore(Type objectType0, CriteriaOperator dbCriteria, string alias) {
            List<object> objects0 = null;
            var preResults = new List<PreResult>();
            var postResults = new List<PostResult>();
            var mapping0 = mappings[objectType0];
            var statement0 = PhaseOne(mapping0, objectType0, dbCriteria, alias);
            preResults.Add(new PreResult() { Mapping = mapping0, ObjectType = objectType0, Statement = statement0 });
            while(preResults.Count > 0) {
                var statements = preResults.Select(p => p.Statement).ToArray();
                var selectedData = dataStore.SelectData(statements);
                var toLoad = new List<PreResult>();
                for(int i = 0; i < selectedData.ResultSet.Length; i++) {
                    var mapping = preResults[i].Mapping;
                    var result = selectedData.ResultSet[i];
                    var objects = PhaseTwo(mapping, preResults[i].ObjectType, result, toLoad);
                    if(objects0 == null) {
                        objects0 = objects;
                    }
                    postResults.Add(new PostResult() { Mapping = mapping, Objects = objects, Result = result });
                }
                preResults = toLoad;
            }
            foreach(var postResult in postResults) {
                PhaseThree(postResult.Mapping, postResult.Objects, postResult.Result);
            }
            return objects0;
        }
        public object LoadObjectByKey(Type objectType, object key) {
            var alias = "T";
            var objects = LoadObjectsCore(objectType, BuildByKeyCriteria(objectType, key, alias), alias);
            if(objects.Count == 1) {
                return objects[0];
            }
            if(objects.Count == 0) {
                return null;
            }
            throw new DataException();
        }
    }

    class DataStoreObjectSaver {
        private IDataStore dataStore;
        private Dictionary<Type, DataStoreMapping> mappings;
        public DataStoreObjectSaver(Dictionary<Type, DataStoreMapping> mappings, IDataStore dataStore) {
            this.mappings = mappings;
            this.dataStore = dataStore;
        }
        public void SaveObjects(ICollection toInsert, ICollection toUpdate, ICollection toDelete) {
            var statements = new List<ModificationStatement>();
            var identityAwaiters = new List<Action<object>>();
            foreach(var obj in toDelete) {
                DeleteObject(obj, statements);
            }
            foreach(var obj in toInsert) {
                InsertObject(obj, statements, identityAwaiters);
            }
            foreach(var obj in toUpdate) {
                UpdateObject(obj, statements);
            }
            var result = dataStore.ModifyData(statements.ToArray());
            foreach(var identity in result.Identities) {
                identityAwaiters[identity.Tag - 1].Invoke(identity.Value);
            }
        }
        private void DeleteObject(object obj, IList<ModificationStatement> statements) {
            if(mappings.TryGetValue(obj.GetType(), out var mapping)) {
                string alias = null;
                var statement = new DeleteStatement(mapping.Table, alias);
                SetupUpdateDeleteStatement(statement, obj, mapping, alias);
                statements.Add(statement);
            }
        }
        private void InsertObject(object obj, IList<ModificationStatement> statements, List<Action<object>> identityAwaiters) {
            if(mappings.TryGetValue(obj.GetType(), out var mapping)) {
                var statement = new InsertStatement(mapping.Table, "T");
                if(mapping.Table.PrimaryKey != null) {
                    foreach(var columnName in mapping.Table.PrimaryKey.Columns) {
                        var column = mapping.Table.GetColumn(columnName);
                        if(column.IsIdentity) {
                            identityAwaiters.Add(v => { mapping.SetKey(obj, v); });
                            statement.IdentityColumn = column.Name;
                            statement.IdentityColumnType = column.ColumnType;
                            statement.IdentityParameter = new ParameterValue(identityAwaiters.Count);
                            break;
                        }
                    }
                }
                SetupInsertUpdateStatement(statement, obj, mapping);
                statements.Add(statement);
            }
        }
        private void UpdateObject(object obj, IList<ModificationStatement> statements) {
            if(mappings.TryGetValue(obj.GetType(), out var mapping)) {
                string alias = null;
                var statement = new UpdateStatement(mapping.Table, alias);
                SetupUpdateDeleteStatement(statement, obj, mapping, alias);
                SetupInsertUpdateStatement(statement, obj, mapping);
                statements.Add(statement);
            }
        }
        private DBColumn GetKeyColumn(DataStoreMapping mapping) {
            return mapping.Table.Columns.First(c => c.IsKey);
        }
        private void SetupUpdateDeleteStatement(ModificationStatement statement, object obj, DataStoreMapping mapping, string alias) {
            statement.Condition = new BinaryOperator(
                new QueryOperand(GetKeyColumn(mapping), alias),
                new OperandValue(mapping.GetKey(obj)),
                BinaryOperatorType.Equal);
        }
        private void SetupInsertUpdateStatement(ModificationStatement statement, object obj, DataStoreMapping mapping) {
            var values = new object[mapping.Table.Columns.Count];
            mapping.Save(obj, values);
            for(int i = 0; i < values.Length; i++) {
                var column = mapping.Table.Columns[i];
                if(!column.IsIdentity) {
                    statement.Operands.Add(new QueryOperand(column, null));
                    statement.Parameters.Add(new OperandValue(values[i]));
                }
            }
        }
    }

    class SimpleDataStoreCriteriaVisitor : ClientCriteriaVisitorBase {
        private DBTable table;
        private string alias;
        public SimpleDataStoreCriteriaVisitor(DBTable table, string alias) {
            this.table = table;
            this.alias = alias;
        }
        protected override CriteriaOperator Visit(OperandProperty theOperand) {
            var column = table.GetColumn(theOperand.PropertyName);
            if(column != null) {
                return new QueryOperand(table.GetColumn(theOperand.PropertyName), alias);
            }
            else {
                return null;
            }
        }
        public static CriteriaOperator Transform(CriteriaOperator criteria, DBTable table, string alias) {
            return new SimpleDataStoreCriteriaVisitor(table, alias).Process(criteria);
        }
    }
}

File: VB/NonPersistentObjectsDemo.Module/DataStoreMapper.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Data
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports DevExpress.Data.Filtering
Imports DevExpress.Data.Filtering.Helpers
Imports DevExpress.Xpo.DB

Namespace NonPersistentObjectsDemo.Module

	Public Class DataStoreMapping
		Public Table As DBTable
		Public Create As Func(Of Object)
		Public Load As Action(Of Object, Object(), ObjectMap)
		Public Save As Action(Of Object, Object())
		Public SetKey As Action(Of Object, Object)
		Public GetKey As Func(Of Object, Object)
		Public RefColumns As IEnumerable(Of Column)
		Public Structure Column
			Public Index As Integer
			Public Type As Type
		End Structure
	End Class

	Friend Class DataStoreObjectLoader
		Private Structure PreResult
			Public Mapping As DataStoreMapping
			Public ObjectType As Type
			Public Statement As SelectStatement
		End Structure
		Private Structure PostResult
			Public Mapping As DataStoreMapping
			Public Objects As List(Of Object)
			Public Result As SelectStatementResult
		End Structure
		Private objectMap As ObjectMap
		Private dataStore As IDataStore
		Private mappings As Dictionary(Of Type, DataStoreMapping)
		Public Sub New(ByVal mappings As Dictionary(Of Type, DataStoreMapping), ByVal dataStore As IDataStore, ByVal objectMap As ObjectMap)
			Me.mappings = mappings
			Me.dataStore = dataStore
			Me.objectMap = objectMap
		End Sub
		Private Function GetKeyColumn(ByVal mapping As DataStoreMapping) As DBColumn
			Return mapping.Table.Columns.First(Function(c) c.IsKey)
		End Function
		Private Function BuildByKeyCriteria(ByVal objectType As Type, ByVal key As Object, ByVal [alias] As String) As CriteriaOperator
			Return New BinaryOperator(New QueryOperand(GetKeyColumn(mappings(objectType)), [alias]), New OperandValue(key), BinaryOperatorType.Equal)
		End Function
		Private Function PhaseOne(ByVal mapping As DataStoreMapping, ByVal objectType As Type, ByVal dbCriteria As CriteriaOperator, ByVal [alias] As String) As SelectStatement
			Dim statement = New SelectStatement(mapping.Table, [alias])
			statement.Condition = dbCriteria
			For Each column In mapping.Table.Columns
				statement.Operands.Add(New QueryOperand(column, [alias]))
			Next column
			Return statement
		End Function
		Private Function PhaseTwo(ByVal mapping As DataStoreMapping, ByVal objectType As Type, ByVal result As SelectStatementResult, ByVal toLoad As List(Of PreResult)) As List(Of Object)
			Dim objects As New List(Of Object)()
			Dim keyColumnIndex As Integer = mapping.Table.Columns.IndexOf(GetKeyColumn(mapping))
			Dim refColumns As List(Of DataStoreMapping.Column) = mapping.RefColumns.ToList()
			For Each row In result.Rows
				Dim key = row.Values(keyColumnIndex)
				If key Is Nothing Then
					Throw New DataException("Key cannot be null.")
				End If
				Dim obj = objectMap.Get(objectType, key)
				If obj Is Nothing Then
					obj = mapping.Create()
					objectMap.Add(objectType, key, obj)
				End If
				objects.Add(obj)
				For Each member In refColumns
					Dim [alias] = "T"
					Dim dbCriteria = BuildByKeyCriteria(member.Type, row.Values(member.Index), [alias])
					Dim memberMapping = mappings(member.Type)
					toLoad.Add(New PreResult() With {
						.Mapping = memberMapping,
						.ObjectType = member.Type,
						.Statement = PhaseOne(memberMapping, member.Type, dbCriteria, [alias])
					})
				Next member
			Next row
			Return objects
		End Function
		Private Sub PhaseThree(ByVal mapping As DataStoreMapping, ByVal objects As List(Of Object), ByVal result As SelectStatementResult)
			For i As Integer = 0 To objects.Count - 1
				mapping.Load(objects(i), result.Rows(i).Values, objectMap)
				objectMap.Accept(objects(i))
			Next i
		End Sub
		Public Function LoadObjects(ByVal objectType As Type, ByVal criteria As CriteriaOperator) As IList(Of Object)
			Dim mapping = mappings(objectType)
			Dim [alias] = "T"
			Dim dbCriteria = SimpleDataStoreCriteriaVisitor.Transform(criteria, mapping.Table, [alias])
			Return LoadObjectsCore(objectType, dbCriteria, [alias])
		End Function
		Private Function LoadObjectsCore(ByVal objectType0 As Type, ByVal dbCriteria As CriteriaOperator, ByVal [alias] As String) As IList(Of Object)
			Dim objects0 As List(Of Object) = Nothing
			Dim preResults = New List(Of PreResult)()
			Dim postResults = New List(Of PostResult)()
			Dim mapping0 = mappings(objectType0)
			Dim statement0 = PhaseOne(mapping0, objectType0, dbCriteria, [alias])
			preResults.Add(New PreResult() With {
				.Mapping = mapping0,
				.ObjectType = objectType0,
				.Statement = statement0
			})
			Do While preResults.Count > 0
				Dim statements = preResults.Select(Function(p) p.Statement).ToArray()
				Dim selectedData = dataStore.SelectData(statements)
				Dim toLoad = New List(Of PreResult)()
				For i As Integer = 0 To selectedData.ResultSet.Length - 1
					Dim mapping = preResults(i).Mapping
					Dim result = selectedData.ResultSet(i)
					Dim objects = PhaseTwo(mapping, preResults(i).ObjectType, result, toLoad)
					If objects0 Is Nothing Then
						objects0 = objects
					End If
					postResults.Add(New PostResult() With {
						.Mapping = mapping,
						.Objects = objects,
						.Result = result
					})
				Next i
				preResults = toLoad
			Loop
			For Each postResult In postResults
				PhaseThree(postResult.Mapping, postResult.Objects, postResult.Result)
			Next postResult
			Return objects0
		End Function
		Public Function LoadObjectByKey(ByVal objectType As Type, ByVal key As Object) As Object
			Dim [alias] = "T"
			Dim objects = LoadObjectsCore(objectType, BuildByKeyCriteria(objectType, key, [alias]), [alias])
			If objects.Count = 1 Then
				Return objects(0)
			End If
			If objects.Count = 0 Then
				Return Nothing
			End If
			Throw New DataException()
		End Function
	End Class

	Friend Class DataStoreObjectSaver
		Private dataStore As IDataStore
		Private mappings As Dictionary(Of Type, DataStoreMapping)
		Public Sub New(ByVal mappings As Dictionary(Of Type, DataStoreMapping), ByVal dataStore As IDataStore)
			Me.mappings = mappings
			Me.dataStore = dataStore
		End Sub
		Public Sub SaveObjects(ByVal toInsert As ICollection, ByVal toUpdate As ICollection, ByVal toDelete As ICollection)
			Dim statements = New List(Of ModificationStatement)()
			Dim identityAwaiters = New List(Of Action(Of Object))()
			For Each obj In toDelete
				DeleteObject(obj, statements)
			Next obj
			For Each obj In toInsert
				InsertObject(obj, statements, identityAwaiters)
			Next obj
			For Each obj In toUpdate
				UpdateObject(obj, statements)
			Next obj
			Dim result = dataStore.ModifyData(statements.ToArray())
			For Each identity In result.Identities
				identityAwaiters(identity.Tag - 1).Invoke(identity.Value)
			Next identity
		End Sub
		Private Sub DeleteObject(ByVal obj As Object, ByVal statements As IList(Of ModificationStatement))
			Dim mapping As DataStoreMapping
			If mappings.TryGetValue(obj.GetType(), mapping) Then
				Dim [alias] As String = Nothing
				Dim statement = New DeleteStatement(mapping.Table, [alias])
				SetupUpdateDeleteStatement(statement, obj, mapping, [alias])
				statements.Add(statement)
			End If
		End Sub
		Private Sub InsertObject(ByVal obj As Object, ByVal statements As IList(Of ModificationStatement), ByVal identityAwaiters As List(Of Action(Of Object)))
			Dim mapping As DataStoreMapping
			If mappings.TryGetValue(obj.GetType(), mapping) Then
				Dim statement = New InsertStatement(mapping.Table, "T")
				If mapping.Table.PrimaryKey IsNot Nothing Then
					For Each columnName In mapping.Table.PrimaryKey.Columns
						Dim column = mapping.Table.GetColumn(columnName)
						If column.IsIdentity Then
							identityAwaiters.Add(Sub(v)
								mapping.SetKey(obj, v)
							End Sub)
							statement.IdentityColumn = column.Name
							statement.IdentityColumnType = column.ColumnType
							statement.IdentityParameter = New ParameterValue(identityAwaiters.Count)
							Exit For
						End If
					Next columnName
				End If
				SetupInsertUpdateStatement(statement, obj, mapping)
				statements.Add(statement)
			End If
		End Sub
		Private Sub UpdateObject(ByVal obj As Object, ByVal statements As IList(Of ModificationStatement))
			Dim mapping As DataStoreMapping
			If mappings.TryGetValue(obj.GetType(), mapping) Then
				Dim [alias] As String = Nothing
				Dim statement = New UpdateStatement(mapping.Table, [alias])
				SetupUpdateDeleteStatement(statement, obj, mapping, [alias])
				SetupInsertUpdateStatement(statement, obj, mapping)
				statements.Add(statement)
			End If
		End Sub
		Private Function GetKeyColumn(ByVal mapping As DataStoreMapping) As DBColumn
			Return mapping.Table.Columns.First(Function(c) c.IsKey)
		End Function
		Private Sub SetupUpdateDeleteStatement(ByVal statement As ModificationStatement, ByVal obj As Object, ByVal mapping As DataStoreMapping, ByVal [alias] As String)
			statement.Condition = New BinaryOperator(New QueryOperand(GetKeyColumn(mapping), [alias]), New OperandValue(mapping.GetKey(obj)), BinaryOperatorType.Equal)
		End Sub
		Private Sub SetupInsertUpdateStatement(ByVal statement As ModificationStatement, ByVal obj As Object, ByVal mapping As DataStoreMapping)
			Dim values = New Object(mapping.Table.Columns.Count - 1){}
			mapping.Save(obj, values)
			For i As Integer = 0 To values.Length - 1
				Dim column = mapping.Table.Columns(i)
				If Not column.IsIdentity Then
					statement.Operands.Add(New QueryOperand(column, Nothing))
					statement.Parameters.Add(New OperandValue(values(i)))
				End If
			Next i
		End Sub
	End Class

	Friend Class SimpleDataStoreCriteriaVisitor
		Inherits ClientCriteriaVisitorBase

		Private table As DBTable
		Private [alias] As String
		Public Sub New(ByVal table As DBTable, ByVal [alias] As String)
			Me.table = table
			Me.alias = [alias]
		End Sub
		Protected Overrides Function Visit(ByVal theOperand As OperandProperty) As CriteriaOperator
			Dim column = table.GetColumn(theOperand.PropertyName)
			If column IsNot Nothing Then
				Return New QueryOperand(table.GetColumn(theOperand.PropertyName), [alias])
			Else
				Return Nothing
			End If
		End Function
		Public Shared Function Transform(ByVal criteria As CriteriaOperator, ByVal table As DBTable, ByVal [alias] As String) As CriteriaOperator
			Return (New SimpleDataStoreCriteriaVisitor(table, [alias])).Process(criteria)
		End Function
	End Class
End Namespace
